docroot/modules/custom/va_gov_backend/src/EventSubscriber/LastEditorSaveEventSubscriber.php

```diff
@@ -73,9 +73,8 @@ public function formAlter(FormAlterEvent $event): void {
       $node = $form_object->getEntity();
 
       $form['field_last_saved_by_an_editor']['#access'] = FALSE;
-      $form['actions']['submit']['#submit'][] = [
-        $this, 'lastSavedByEditorSetTimestamp',
-      ];
+      array_unshift($form['actions']['submit']['#submit'],
+        [$this, 'lastSavedByEditorSetTimestamp']);
```

Comment:
I am a little confused on the format here.  Why are we adding $this to the array of the submit handler that should  just be a callback name.
What does passing it $this do?  I can't find that pattern anywhere.
```suggestion
      array_unshift($form['actions']['submit']['#submit'], 'lastSavedByEditorSetTimestamp');
```

Reply:
That's [this pattern](https://www.php.net/manual/en/language.types.callable.php):

![Screen Shot 2022-11-25 at 8 16 38 AM](https://user-images.githubusercontent.com/1318579/203993713-7bdde232-ac61-48c3-afa5-eaf1b2aefc08.png)

Reply:
@swirtSJW this is how you add the callback inside event subscribers. Otherwise, that method, in this case `lastSavedByEditorSetTimestamp`, would have to go in the `.module` file.

Reply:
Oh thanks for explaining. I see why you did that.  That seems like a fair amount of overhead to pass that instantiated object around the steps of a form process, just to be able to call a  4 line function.   A less intensive way to get at it, without moving that function to the .module, would be to change it to a public static function then call it using its namespaced path
like this

array_unshift($form['actions']['submit']['#submit'], 'Drupal\va_gov_backend\EventSubscriber\LastEditorSaveEventSubscriber::lastSavedByEditorSetTimestamp');

Granted, this particular object is pretty small and not a lot of overhead, but I wouldn't want somebody finding and following this pattern, when they have an object that is more gigantic.

Reply:
I made that function public and used the static call I referenced above and it works fine.

Reply:
I'd be surprised if there were meaningful overhead to the `$this` approach for an object of any size, since it should just be passing a reference around.

Reply:
Just pushed this change.

Reply:
@ndouglas you are right it would be by reference so there should be no additional overhead IF that event subscriber instance still exists.  I am not certain when event subscriber instances get cleaned up.  Maybe they never do (spine shivers).  In my head I was not picturing it as existing past the first load of node edit, but formAlter does fire on the submit too, so maybe it does persist, or maybe a new instance does?.  

I think in terms of architecture though... calling a static method seems a safer implementation.  
**Example from a distopian future:**  Somebody adds another submit handler to this class, and both submit handlers pass in the same instance in the callback.  However the new callback since it is using public method rather than static, alters the object in some way that the original callback does not expect.  The one callback could alter the outcome of the other, because they both have the ability/risk to interact with the same instance of the object.  

By using only static calls, we prevent one from inadvertently interacting with the other and all is right with the world  :)

Reply:
Now I feel like you're just arguing against OOP 😛  You're not gonna get much argument from me there.  

But in many cases (most?), we _will_ need to consult services from the service container, and in those cases I think it's better to perform calls on an instantiated object with the services DI'ed than from a static method or function that calls `\Drupal::whatever()`.  That's the sort of situation where I'm worried about setting a precedent.